### DISCUSSION TITLE
Timeout CXF

### DISCUSSION THREAD

**Discussion by luisfalero** [quarkiverse/quarkus-cxf#1282] [General]
Hi Everyone,

Could someone help me by indicating how I could add timeout to how much an external SOAP service consumes?

I tried to add the commented lines but it shows an error when adding: "Client client = ClientProxy.getClient(result.createClient());"

```java
@Produces
@SessionScoped
@Named("calculatorSoapEndpoint")
CxfEndpoint calculatorSoapEndpoint() throws Exception {
    final CxfEndpoint result = new CxfEndpoint();
    result.setServiceClass(CalculatorSoap.class);
    result.setContinuationTimeout(1);
    result.setPollingConsumerBlockTimeout(1);        
    result.setAddress("http://www.dneonline.com/calculator.asmx");
    result.setWsdlURL("wsdl/CalculatorService.wsdl");
        
    // Configuración del cliente CXF
    //Client client = ClientProxy.getClient(result.createClient());
    //HTTPConduit httpConduit = (HTTPConduit) client.getConduit();
    //HTTPClientPolicy httpClientPolicy = new HTTPClientPolicy();
    //httpClientPolicy.setConnectionTimeout(5000); // Tiempo de espera de conexión en milisegundos
    //httpClientPolicy.setReceiveTimeout(5000);    // Tiempo de espera para recibir en milisegundos
    //httpClientPolicy.setAllowChunking(false);    // Deshabilitar el uso de transferencia de datos fragmentados
    //httpConduit.setClient(httpClientPolicy);
    return result;
}
```

**Comment by ppalaga** [comment]
The Camel way would be to have an implementation of `org.apache.camel.component.cxf.jaxws.CxfConfigurer` where you can configure the client and conduit:

```java
class MyCxfConfigurer implements CxfConfigurer {
...
    @Override
    public void configureClient(Client client) {
        HTTPConduit httpConduit = (HTTPConduit) client.getConduit();
        HTTPClientPolicy httpClientPolicy = new HTTPClientPolicy();
        httpClientPolicy.setConnectionTimeout(5000); // Tiempo de espera de conexión en milisegundos
        httpClientPolicy.setReceiveTimeout(5000);    // Tiempo de espera para recibir en milisegundos
        httpClientPolicy.setAllowChunking(false);    // Deshabilitar el uso de transferencia de datos fragmentados
        httpConduit.setClient(httpClientPolicy);
        setupHttpConduit(httpConduit);
    }

}
```

you can then set the configurer on CxfEndpoint.

The Quarkus CXF way (for clients injected through `@CXFClient`) would be to do it in application.properties using the various `*-timeout ` options: https://docs.quarkiverse.io/quarkus-cxf/dev/reference/extensions/quarkus-cxf.html#quarkus-cxf_quarkus-cxf-client-clients-connection-timeout

**Reply by luisfalero** [reply]
Hi @ppalaga 

How would I call the MyCxfConfigurer class from the class that has the calculatorSoapEndpoint() method?

I mean, what value can I put in the input parameter of the configureClient() method?

**Reply by luisfalero** [reply]
My complete JAVA class:

@ApplicationScoped
public class CxfSoapClientRoutes extends RouteBuilder {
       
    @Override
    public void configure() throws Exception {
        rest()
            .post("/add")
            .routeId("RouteRest")
            .consumes("application/json")
            .produces("application/json")
            .to("direct:consumeSoapService");

        from("direct:consumeSoapService")
            .routeId("RouteSoap")
            .to("cxf:bean:calculatorSoapEndpoint?dataFormat=POJO")          
    }

    @Produces
    @SessionScoped
    @Named("calculatorSoapEndpoint")
    CxfEndpoint calculatorSoapEndpoint() throws Exception {
        final CxfEndpoint result = new CxfEndpoint();
        result.setServiceClass(CalculatorSoap.class);
        result.setContinuationTimeout(1);
        result.setPollingConsumerBlockTimeout(1);        
        result.setAddress("http://www.dneonline.com/calculator.asmx");
        result.setWsdlURL("wsdl/CalculatorService.wsdl");
        
        // Configuración del cliente CXF
        //Client client = ClientProxy.getClient(result.createClient());
        //HTTPConduit httpConduit = (HTTPConduit) client.getConduit();
        //HTTPClientPolicy httpClientPolicy = new HTTPClientPolicy();
        //httpClientPolicy.setConnectionTimeout(5000); // Tiempo de espera de conexión en milisegundos
        //httpClientPolicy.setReceiveTimeout(5000);    // Tiempo de espera para recibir en milisegundos
        //httpClientPolicy.setAllowChunking(false);    // Deshabilitar el uso de transferencia de datos fragmentados
        //httpConduit.setClient(httpClientPolicy);     
        return result;
    }
}

**Comment by ppalaga** [comment]
You do not need to call it, just pass an instance:

```java
@Produces
@SessionScoped
@Named("calculatorSoapEndpoint")
CxfEndpoint calculatorSoapEndpoint() throws Exception {
    final CxfEndpoint result = new CxfEndpoint();
 
     ...
     result.setCxfConfigurer(new MyCxfConfigurer());
     ...
     return result;
}
```
